File: 2019/python/day1.py

with open("../data/data1.txt") as f:
    data = f.readlines()
    data = [int(x) for x in data]


def fuel_requirement(fuel):
    return max(fuel // 3 - 2, 0)


def fuel_requirement_recursive(fuel):
    need = max(fuel // 3 - 2, 0)
    if need == 0:
        return need
    else:
        return need + fuel_requirement_recursive(need)


# Part 1 ----
print(sum(map(fuel_requirement, data)))

# Part 2 ----
print(sum(map(fuel_requirement_recursive, data)))

File: 2019/R/day1.R

library(magrittr)
library(testthat)

aa <- read.delim(
  "../data/data1.txt",
  header = FALSE,
  colClasses = "numeric",
  col.names = "module"
)


get_fuel_requirement <- function(x) { pmax(floor(x/3) - 2, 0) }

get_fuel_requirement_recursive <- function(x) {
  fuel <- numeric(length(x))
  while (any(x > 6)) {
    x <- sapply(x, get_fuel_requirement)
    fuel <- fuel + x
  }
  fuel
}

get_fuel_sum <- function(x) {
  x[["module"]] %>% get_fuel_requirement() %>% sum()
}

get_fuel_sum_recursive <- function(x) {
  x[["module"]] %>% get_fuel_requirement_recursive() %>% sum()
}

test_that("Unit Tests Part 1", {
  expect_equal(get_fuel_requirement(12), 2)
  expect_equal(get_fuel_requirement(14), 2)
  expect_equal(get_fuel_requirement(1969), 654)
  expect_equal(get_fuel_requirement(100756), 33583)
})

test_that("Unit Tests Part 2", {
  expect_equal(get_fuel_requirement_recursive(14), 2)
  expect_equal(get_fuel_requirement_recursive(1969), 966)
  expect_equal(get_fuel_requirement_recursive(100756), 50346)

  expect_equal(get_fuel_requirement_recursive(c(14,1969,100756)), c(2,966,50346))
})

get_fuel_sum(aa)
get_fuel_sum_recursive(aa)
